core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala

```diff
@@ -0,0 +1,219 @@
+        val ref = childFactory(context)
+        pool.update(ref, WorkerData(NoData(), Free))
+        ref
+    }
+
+    /** Creates a new prewarmed container */
+    def prewarmContainer() = prewarmConfig.foreach(config => childFactory(context) ! Start(config.exec, config.memoryLimit))
+
+    /**
+     * Takes a prewarm container out of the prewarmed pool
+     * iff a container with a matching kind is found.
+     *
+     * @param kind the kind you want to invoke
+     * @return the container iff found
+     */
+    def takePrewarmContainer(action: ExecutableWhiskAction) = prewarmConfig.flatMap { config =>
+        val kind = action.exec.kind
+        val memory = action.limits.memory.megabytes.MB
+        prewarmedPool.find {
+            case (_, WorkerData(PreWarmedData(_, `kind`, `memory`), _)) => true
+            case _ => false
+        }.map {
+            case (ref, data) =>
+                // Move the container to the usual pool
+                pool.update(ref, data)
+                prewarmedPool.remove(ref)
+                // Create a new prewarm container
+                prewarmContainer()
+
+                ref
+        }
+    }
+
+    /** Removes a container and updates state accordingly. */
+    def removeContainer(toDelete: ActorRef) = {
+        toDelete ! Remove
+        pool.remove(toDelete)
+    }
+}
+
+object ContainerPool {
+    /**
+     * Finds the best container for a given job to run on.
+     *
+     * Selects an arbitrary warm container from the passed pool of idle containers
+     * that matches the action and the invocation namespace. The implementation uses
+     * matching such that structural equality of action and the invocation namespace
+     * is required.
+     * Returns None iff no matching container is in the idle pool.
+     * Does not consider pre-warmed containers.
+     *
+     * @param action the action to run
+     * @param namespace the namespace, that wants to run the action
+     * @param idles a map of idle containers, awaiting work
+     * @return a container if one found
+     */
+    def schedule[A](action: ExecutableWhiskAction, namespace: EntityName, idles: Map[A, WorkerData]): Option[A] = {
+        idles.find {
+            case (_, WorkerData(WarmedData(_, `namespace`, `action`, _), Free)) => true
+            case _ => false
+        }.map(_._1)
+    }
+
+    /**
+     * Finds the best container to remove to make space for the job passed to run.
+     *
+     * Determines which namespace consumes most resources in the current pool and
+     * takes away one of their containers iff the namespace placing the new job is
+     * not already the most consuming one.
+     *
+     * @param namespace the namespace that wants to get a container
+     * @param pool a map of all containers in the pool
+     * @return a container to be removed iff found
+     */
+    def remove[A](namespace: EntityName, pool: Map[A, WorkerData]): Option[A] = {
+        val grouped = pool.collect {
```

Comment:
I agree, this is specifically scoped to the rather small pools we're having now. When we expand the idea we'll have to revisit those.